core/src/main/java/org/apache/calcite/rel/core/JoinInfo.java

```diff
@@ -50,30 +52,59 @@ public class JoinInfo {
 
   /** Creates a JoinInfo. */
   protected JoinInfo(ImmutableIntList leftKeys, ImmutableIntList rightKeys,
-      ImmutableList<RexNode> nonEquiConditions) {
-    this.leftKeys = requireNonNull(leftKeys, "leftKeys");
-    this.rightKeys = requireNonNull(rightKeys, "rightKeys");
-    this.nonEquiConditions =
-        requireNonNull(nonEquiConditions, "nonEquiConditions");
+      ImmutableList<RexNode> nonEquiConditions, ImmutableList<Boolean> filterNulls,
```

Comment:
changing the signature of protected methods is a breaking change, which will affect projects that use calcite.
I don't understand why JoinInfo needs to be changed if the problem is in the EnumerableJoinRule.

Reply:
What is the problem of joins and null handling?
Joins can use arbitrary predicates.
What is an "equality condition"?

Reply:
Maybe you can the details in jira 6904. leftKeys and rightKeys in JoinInfo contain equi condition.

Reply:
CALCITE-6904 does not say anything about standard joins, it says there is a problem with EnumerableJoinRule

Reply:
In EnumerableJoinRule, JoinInfo is used to split the Equi and non-equi conditions, and then the two are combined as the join condition. And then will use HashJoin. If it does not do such a split, it is fine to use NestedLoopJoin directly with the original join condition.

Reply:
Maybe modify RelOptUtil.splitJoinCondition is a right way?

Reply:
Maybe the question boils down at how HashJoin is supposed to be implemented.
In our implementation when the condition is equality (=) we filter out nulls prior to the join.
If the condition is `IS NOT DISTINCT FROM` we don't filter the nulls.
You can use a HashJoin in both cases.

Reply:
So you mean that it is reasonable for us to modify the execution logic of hashjoin? You said before that we should not change this.

Reply:
I am not sure what "the execution logic of hashjoin" is. I am not sure where I made that statement, I could not find it in the JIRA case or this conversation. 

There must be some conversion rules in the Enumerable implementation which decide when and how to implement a LogicalJoin using a HashJoin - I haven't looked at that part of the code. Maybe you are suggesting that the logic in that transformation is incorrect or inefficient.

Reply:
> Maybe the question boils down at how HashJoin is supposed to be implemented. In our implementation when the condition is equality (=) we filter out nulls prior to the join. If the condition is `IS NOT DISTINCT FROM` we don't filter the nulls. You can use a HashJoin in both cases.

I mean, do we need to modify the logic here?

Reply:
I am tempted to say "no" - JoinInfo is in the rel/core package, and has nothing to do with the enumerable implementation. I think any changes belong to that package.

Reply:
I wonder if it would be better to treat IS NOT DISTINCT FROM as a non-equi condition and use NestedLoopJoin to execute it? I think it should be possible to implement it by modifying RelOptUtil.splitJoinCondition without modifying JoinInfo.

Reply:
There are lots of tests using joins and Enumerable which work.
What does not work?

Reply:
Maybe here https://github.com/apache/calcite/blob/c72e267f9c1b94abb3f4a7e31d14f7b4ed676ce1/core/src/main/java/org/apache/calcite/plan/RelOptUtil.java#L1697